File: backend/engine/tm_engine.py
from typing import List, Dict
try:
    from models.schemas import TMMachine, SimulationResult, SimulationStepTrace
except ImportError:
    from ..models.schemas import TMMachine, SimulationResult, SimulationStepTrace

def simulate_tm(machine: TMMachine, input_str: str, max_steps: int = 1000) -> SimulationResult:
    blank = machine.blankSymbol or '_'
    state_lookup = {s.id: s.label for s in machine.states}

    tape_dict: Dict[int, str] = {}
    for i, char in enumerate(input_str):
        tape_dict[i] = char

    head_pos = 0
    current_state = machine.startState
    traces: List[SimulationStepTrace] = []

    def get_tape_window():
        min_idx = min(min(tape_dict.keys(), default=0), head_pos)
        max_idx = max(max(tape_dict.keys(), default=0), head_pos)
        start = min(0, min_idx - 2)
        end = max(len(input_str) + 2, max_idx + 3)
        window = [tape_dict.get(i, blank) for i in range(start, end)]
        rel_head = head_pos - start
        return window, rel_head

    # Initial Step
    init_window, init_head = get_tape_window()
    is_init_accept = current_state in machine.acceptStates
    is_init_reject = current_state in (machine.rejectStates or [])

    traces.append(
        SimulationStepTrace(
            stepIndex=0,
            activeStates=[current_state],
            remainingInput=input_str,
            readSymbol=None,
            activeTransitionId=None,
            tape=init_window,
            headIndex=init_head,
            actionSummary=f"TM Initialized at '{state_lookup.get(current_state, current_state)}', Head pos: {head_pos}",
            status="HALTED_ACCEPT" if is_init_accept else ("HALTED_REJECT" if is_init_reject else "RUNNING"),
        )
    )

    if is_init_accept or is_init_reject:
        return SimulationResult(
            accepted=is_init_accept,
            finalStatus="ACCEPTED" if is_init_accept else "REJECTED",
            totalSteps=1,
            traces=traces,
            message=f"TM halted immediately in {'accept' if is_init_accept else 'reject'} state.",
        )

    steps_count = 0
    while steps_count < max_steps:
        steps_count += 1
        current_sym = tape_dict.get(head_pos, blank)

        if current_state in machine.acceptStates:
            w, h = get_tape_window()
            traces.append(
                SimulationStepTrace(
                    stepIndex=len(traces),
                    activeStates=[current_state],
                    remainingInput="",
                    readSymbol=current_sym,
                    activeTransitionId=None,
                    tape=w,
                    headIndex=h,
                    actionSummary=f"TM reached ACCEPT state '{state_lookup.get(current_state, current_state)}'. Halting.",
                    status="ACCEPTED",
                )
            )
            return SimulationResult(
                accepted=True,
                finalStatus="ACCEPTED",
                totalSteps=len(traces),
                traces=traces,
                message=f"Accepted: TM halted in accept state '{state_lookup.get(current_state, current_state)}'.",
            )

        if current_state in (machine.rejectStates or []):
            w, h = get_tape_window()
            traces.append(
                SimulationStepTrace(
                    stepIndex=len(traces),
                    activeStates=[current_state],
                    remainingInput="",
                    readSymbol=current_sym,
                    activeTransitionId=None,
                    tape=w,
                    headIndex=h,
                    actionSummary=f"TM reached REJECT state '{state_lookup.get(current_state, current_state)}'. Halting.",
                    status="REJECTED",
                )
            )
            return SimulationResult(
                accepted=False,
                finalStatus="REJECTED",
                totalSteps=len(traces),
                traces=traces,
                message=f"Rejected: TM halted in reject state '{state_lookup.get(current_state, current_state)}'.",
            )

        # Lookup transition: (current_state, current_sym)
        transition = next(
            (t for t in machine.transitions if t.from_state == current_state and t.readSymbol == current_sym),
            None
        )

        if transition is None:
            w, h = get_tape_window()
            traces.append(
                SimulationStepTrace(
                    stepIndex=len(traces),
                    activeStates=[current_state],
                    remainingInput="",
                    readSymbol=current_sym,
                    activeTransitionId=None,
                    tape=w,
                    headIndex=h,
                    actionSummary=f"No transition for ('{state_lookup.get(current_state, current_state)}', '{current_sym}'). TM crashed / halted implicitly.",
                    status="REJECTED",
                )
            )
            return SimulationResult(
                accepted=False,
                finalStatus="REJECTED",
                totalSteps=len(traces),
                traces=traces,
                message=f"Rejected: No transition for ('{state_lookup.get(current_state, current_state)}', '{current_sym}').",
            )

        # Execute transition: Write symbol & Move head
        tape_dict[head_pos] = transition.writeSymbol
        old_state = current_state
        current_state = transition.to_state

        if transition.direction == 'R':
            head_pos += 1
        elif transition.direction == 'L':
            head_pos -= 1

        w, h = get_tape_window()
        is_acc = current_state in machine.acceptStates
        is_rej = current_state in (machine.rejectStates or [])

        traces.append(
            SimulationStepTrace(
                stepIndex=len(traces),
                activeStates=[current_state],
                remainingInput="",
                readSymbol=current_sym,
                activeTransitionId=transition.id,
                tape=w,
                headIndex=h,
                actionSummary=f"Read '{current_sym}', wrote '{transition.writeSymbol}', moved {transition.direction} -> '{state_lookup.get(current_state, current_state)}'",
                status=("ACCEPTED" if is_acc else "REJECTED") if (is_acc or is_rej) else "RUNNING",
            )
        )

        if is_acc:
            return SimulationResult(
                accepted=True,
                finalStatus="ACCEPTED",
                totalSteps=len(traces),
                traces=traces,
                message=f"Accepted: Reached accept state '{state_lookup.get(current_state, current_state)}'.",
            )
        if is_rej:
            return SimulationResult(
                accepted=False,
                finalStatus="REJECTED",
                totalSteps=len(traces),
                traces=traces,
                message=f"Rejected: Reached explicit reject state.",
            )

    return SimulationResult(
        accepted=False,
        finalStatus="REJECTED",
        totalSteps=len(traces),
        traces=traces,
        message=f"Halted: Maximum step limit ({max_steps}) exceeded.",
    )

File: frontend/src/engine/tmEngine.ts
import { BLANK, SimulationResult, SimulationStepTrace, TMMachine } from '../types/automata';

export function isBlank(sym: string, blankSymbol: string): boolean {
  if (sym === undefined || sym === null || sym === '') return true;
  const targetBlank = blankSymbol || '_';
  return sym === targetBlank || sym === '_' || sym === '␣';
}

export function normalizeTapeSymbol(sym: string, blankSymbol: string): string {
  if (isBlank(sym, blankSymbol)) return blankSymbol || '_';
  return sym;
}

export function simulateTM(rawMachine: TMMachine, inputString: string, maxSteps = 1000): SimulationResult {
  const machine = {
    ...rawMachine,
    states: rawMachine.states || [],
    transitions: rawMachine.transitions || [],
    acceptStates: rawMachine.acceptStates || [],
    rejectStates: rawMachine.rejectStates || [],
    inputAlphabet: rawMachine.inputAlphabet || [],
    tapeAlphabet: rawMachine.tapeAlphabet || [],
  };
  const blank = machine.blankSymbol || '_';
  const startState = machine.startState;
  const traces: SimulationStepTrace[] = [];

  if (!startState) {
    return {
      accepted: false,
      finalStatus: 'ERROR',
      traces: [{
        stepIndex: 0,
        activeStates: [],
        remainingInput: inputString,
        inputIndex: 0,
        tape: [blank],
        headIndex: 0,
        actionSummary: 'No start state defined for Turing Machine',
        status: 'ERROR',
      }],
      totalSteps: 1,
      message: 'No start state defined for Turing Machine.',
    };
  }

  // Initialize tape: input symbols + padding blanks on both sides for visual clarity
  const inputChars = inputString.length > 0 ? inputString.split('') : [blank];
  const tape: string[] = [...inputChars];
  let headIndex = 0; // head starts at position 0
  let virtualOffset = 0;
  let currentState = startState;

  // Add initial trace
  traces.push({
    stepIndex: 0,
    activeStates: [currentState],
    currentSymbol: tape[headIndex] || blank,
    inputIndex: 0,
    remainingInput: inputString,
    tape: [...tape],
    headIndex: headIndex,
    tapeOffset: virtualOffset,
    actionSummary: `Initial Configuration: State ${currentState}, Head at index 0 reading '${tape[headIndex]}'`,
    status: 'RUNNING',
  });

  // Check if start state is already accept state
  if (machine.acceptStates.includes(currentState)) {
    traces[0].status = 'HALTED_ACCEPT';
    traces[0].actionSummary += ' (Already in Accept State)';
    return {
      accepted: true,
      finalStatus: 'HALTED_ACCEPT',
      traces,
      totalSteps: 1,
      message: `Turing machine immediately accepted in start state ${currentState}.`,
    };
  }

  for (let step = 1; step <= maxSteps; step++) {
    const currentRead = tape[headIndex] !== undefined ? tape[headIndex] : blank;

    // Find matching transition
    const transition = machine.transitions.find(t => {
      if (t.from !== currentState) return false;
      
      const tRead = t.readSymbol;
      if (isBlank(tRead, blank) && isBlank(currentRead, blank)) return true;
      return tRead === currentRead;
    });

    if (!transition) {
      // No transition -> Machine halts
      const isAccept = machine.acceptStates.includes(currentState);
      const haltStatus = isAccept ? 'HALTED_ACCEPT' : 'HALTED_REJECT';

      traces.push({
        stepIndex: step,
        activeStates: [currentState],
        currentSymbol: currentRead,
        inputIndex: headIndex,
        remainingInput: '',
        tape: [...tape],
        headIndex: headIndex,
        tapeOffset: virtualOffset,
        actionSummary: `No transition found for (State: ${currentState}, Symbol: '${currentRead}'). Halted (${isAccept ? 'ACCEPT' : 'REJECT'}).`,
        status: haltStatus,
      });

      return {
        accepted: isAccept,
        finalStatus: haltStatus,
        traces,
        totalSteps: traces.length,
        message: isAccept
          ? `Turing machine halted in accept state ${currentState}. Final tape: "${tape.join('')}"`
          : `Turing machine halted in non-accepting state ${currentState}. Final tape: "${tape.join('')}"`,
      };
    }

    // Write symbol
    const writeSym = normalizeTapeSymbol(transition.writeSymbol, blank);
    tape[headIndex] = writeSym;

    // Direction move
    let dirDesc = 'Stay';
    if (transition.direction === 'L') {
      dirDesc = 'Move Left';
      if (headIndex === 0) {
        tape.unshift(blank);
        virtualOffset -= 1;
        // headIndex remains 0
      } else {
        headIndex -= 1;
      }
    } else if (transition.direction === 'R') {
      dirDesc = 'Move Right';
      headIndex += 1;
      if (headIndex >= tape.length) {
        tape.push(blank);
      }
    }

    currentState = transition.to;
    const nextRead = tape[headIndex] !== undefined ? tape[headIndex] : blank;

    const isAccept = machine.acceptStates.includes(currentState);
    const isReject = (machine.rejectStates || []).includes(currentState);

    let stepStatus: SimulationStepTrace['status'] = 'RUNNING';
    if (isAccept) stepStatus = 'HALTED_ACCEPT';
    else if (isReject) stepStatus = 'HALTED_REJECT';

    const actionDesc = `Read '${currentRead}' → Write '${writeSym}', ${dirDesc} → State ${currentState}`;

    traces.push({
      stepIndex: step,
      activeStates: [currentState],
      currentSymbol: nextRead,
      inputIndex: headIndex,
      remainingInput: '',
      tape: [...tape],
      headIndex: headIndex,
      tapeOffset: virtualOffset,
      activeTransitionId: transition.id,
      actionSummary: actionDesc,
      status: stepStatus,
    });

    if (isAccept) {
      return {
        accepted: true,
        finalStatus: 'HALTED_ACCEPT',
        traces,
        totalSteps: traces.length,
        message: `Turing machine reached ACCEPT state ${currentState}. Final tape: "${tape.join('')}"`,
      };
    }

    if (isReject) {
      return {
        accepted: false,
        finalStatus: 'HALTED_REJECT',
        traces,
        totalSteps: traces.length,
        message: `Turing machine reached REJECT state ${currentState}. Final tape: "${tape.join('')}"`,
      };
    }
  }

  // Max steps exceeded -> Loop detected
  const lastTrace = traces[traces.length - 1];
  lastTrace.status = 'HALTED_REJECT';
  lastTrace.actionSummary += ' | [Max Step Limit Exceeded / Possible Loop]';

  return {
    accepted: false,
    finalStatus: 'TIMEOUT',
    traces,
    totalSteps: traces.length,
    message: `Simulation reached maximum step limit (${maxSteps}) and was terminated. This may indicate an infinite loop.`,
  };
}
